core/win32/callback.c

```diff
@@ -5647,6 +5647,66 @@ intercept_exception(app_state_at_intercept_t *state)
              */
             mcontext_to_context(cxt, &mcontext, false /* !set_cur_seg */);
 
+            /*
+             * i#2144 : we check if this is a single step exception
+             * where we diverted the address.
+             */
+            if (pExcptRec->ExceptionCode == EXCEPTION_SINGLE_STEP) {
+                instr_t instr;
+
+                instr_init(dcontext, &instr);
+                decode(dcontext, pExcptRec->ExceptionAddress, &instr);
+                /* Checks that exception address translate on a popf. */
+                if (instr_get_opcode(&instr) == OP_popf || instr_get_opcode(&instr) == OP_iret) {
+                    instr_reset(dcontext, &instr);
+                    decode(dcontext, faulting_pc, &instr);
+                    /* Checks that exception was generated by a nop. */
+                    if (instr_get_opcode(&instr) == OP_nop) {
+                        LOG(THREAD, LOG_ASYNCH, 2,
+                            "Caught generated single step exception at "PFX"\n",
+                            pExcptRec->ExceptionAddress);
+                        /* Will continue after one byte popf or iret. */
+                        if (instr_get_opcode(&instr) == OP_popf) {
+                            dcontext->next_tag = mcontext.pc + POPF_LENGTH;
+                        }
+                        else {
+                            dcontext->next_tag = mcontext.pc + IRET_LENGTH;
+                        }
+                        /* Deletes fragment starting at single step exception
+                         * if it exists.
+                         */
+                        f = fragment_lookup_same_sharing(dcontext, dcontext->next_tag,
+                                                         0/*FRAG_PRIVATE*/);
+                        if (f) {
+                            fragment_delete(dcontext, f, FRAGDEL_ALL);
+                        }
+                        /* Sets a field so that build_bb_ilist knows when to stop. */
+                        dcontext->forged_exception_addr = dcontext->next_tag;
+                        /* Will return to execute instruction
+                         * where single step exception will be forged.
+                         */
+                        dcontext->whereami = WHERE_FCACHE;
+                        set_last_exit(dcontext, (linkstub_t *)get_asynch_linkstub());
+                        if (instr_get_opcode(&instr) == OP_iret) {
+                            /* Emulating the rest of iret which is just a pop into rsp
+                             * for x64 and nothing for x86.
+                             */
+#ifdef X64
+                            /* FIXME i#2144 : handle if the pop rsp fails */
+                            state->mc.xsp = *((byte **)state->mc.xsp);
```

Comment:
Is this the right way to emulate the pop rsp ?

About faulting, I see that we should first test readability of this stack.
But what should we do then ?  How can we unroll all iret and generate the correct read exception ?